File: GoLang/solution/solution0417.go
package solution

import "fmt"

func pacificAtlantic(heights [][]int) [][]int {
	visited := make([][]bool, len(heights))
	for i := 0; i < len(heights); i++ {
		visited[i] = make([]bool, len(heights[0]))
	}
	res := make([][]int, 0)
	var p, a bool
	nextMove := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	for i := 0; i < len(heights); i++ {
		for j := 0; j < len(heights[0]); j++ {
			p = false
			a = false
			dfs0417(heights, res, i, j, &p, &a, nextMove, visited)
			if a && p {
				res = append(res, []int{i, j})
			}
			for i := 0; i < len(visited); i++ {
				for j := 0; j < len(visited[0]); j++ {
					visited[i][j] = false
				}
			}
		}
	}
	return res
}

func dfs0417(heights [][]int, res [][]int, currentI int, currentJ int, p *bool, a *bool, nextMove [][]int, visited [][]bool) {
	visited[currentI][currentJ] = true
	//流进了太平洋
	if currentI == 0 || currentJ == 0 {
		*p = true
	}

	//流进了大西洋
	if currentI == len(heights)-1 || currentJ == len(heights[0])-1 {
		*a = true
	}

	for i := 0; i < len(nextMove); i++ {
		nextI := currentI + nextMove[i][0]
		nextJ := currentJ + nextMove[i][1]
		if nextI >= 0 && nextI <= len(heights)-1 && nextJ >= 0 && nextJ <= len(heights[0])-1 {
			if !visited[nextI][nextJ] && heights[nextI][nextJ] <= heights[currentI][currentJ] {
				dfs0417(heights, res, nextI, nextJ, p, a, nextMove, visited)
			}
		}
	}
}

func Test0417() {
	heights := [][]int{{1, 2, 2, 3, 5}, {3, 2, 3, 4, 4}, {2, 4, 5, 3, 1}, {6, 7, 1, 4, 5}, {5, 1, 1, 2, 4}}
	//heights := [][]int{{2, 1}, {1, 2}}
	fmt.Println(pacificAtlantic(heights))
}

File: JavaR2/src/Solution0417.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * <pre>
 * Modify Information:
 * Author       Date          Description
 * ============ ============= ============================
 * maximus         2024/2/27      create
 */
public class Solution0417 {
	int[][] nextMoves = new int[][] {{1,0}, {-1,0}, {0,-1}, {0,1}};

	public List<List<Integer>> pacificAtlantic(int[][] heights) {
		int m = heights.length, n = heights[0].length;

		//count初始化为0,可以流向大西洋+1,可以流向太平洋再+1,等于2的则为节点
		int[][] counts = new int[m][n];
		boolean[][] visited;

		//从边界开始向里递归
		for (int i = 0; i < m; i++) {
			for (int j = 0; j < n; j++) {
				//太平洋流入
				if (i == 0 || j == 0) {
					//一个节点可能从不同的路径流向太平洋，每次都是1条新的路径，因此重置visited数组
					visited = new boolean[m][n];
					dfs1(i, j, -1, -1, heights, counts, visited, true);

				}

				//大西洋流入
				if (i == m - 1 || j == n - 1) {
					visited = new boolean[m][n];
					dfs1(i, j, -1, -1, heights, counts, visited, false);
				}
			}
		}

		List<List<Integer>> res = new ArrayList<>();
		for (int i = 0; i < m; i++) {
			for (int j = 0; j < n; j++) {
				if (counts[i][j] == 2) {
					res.add(Arrays.asList(new Integer[] {i, j}));
				}
			}
		}

		return res;
	}

	private void dfs1(int i, int j, int oldI, int oldJ, int[][] heights, int[][] counts, boolean[][] visited, boolean isPacific) {
		if (i < 0 || i >= heights.length || j < 0 || j >= heights[0].length) {
			return;
		}

		//已经访问过的节点不再访问
		if (visited[i][j]) {
			return;
		}

		visited[i][j] = true;

		//oldI==-1 && oldJ==-1表示在边界上，一定满足
		if (oldI == -1 && oldJ == -1) {
			if (isPacific) {
				if (counts[i][j] == 0) {
					counts[i][j]++;
				}
			} else {
				if (counts[i][j] == 1) {
					counts[i][j]++;
				}
			}
		} else {
			//否则要比较大小
			if (heights[i][j] >= heights[oldI][oldJ]) {
				if (isPacific) {
					if (counts[i][j] == 0) {
						counts[i][j]++;
					}
				} else {
					if (counts[i][j] == 1) {
						counts[i][j]++;
					}
				}
			} else {
				return;
			}
		}

		for (int[] nextMove : nextMoves) {
			int nextI = i + nextMove[0];
			int nextJ = j + nextMove[1];

			dfs1(nextI, nextJ, i, j, heights, counts, visited, isPacific);

		}
		return;
	}
}
